### PR TITLE
Added switch to global enable geocoding

### PR DESCRIPTION
**Pull request by bdm310**
PR I started a while ago and just got around to finishing. Noticed that it addresses #794.

Adds a "Geocoding" boolean switch to Advanced settings to disable the geocoder, no matter which one is selected.

I did not add translations for the settings strings as I'm not a native speaker of anything but English. I'm happy to try and add if necessary.

### CODE REVIEW

File: project/app/src/main/java/org/owntracks/android/support/Preferences.java [hunk 1]
```diff
@@ -950,6 +960,7 @@ public void setObjectBoxMigrated() {
         public static final String DEBUG_LOG                        = "debugLog";
         public static final String DONT_REUSE_HTTP_CLIENT           = "dontReuseHttpClient";
         public static final String FUSED_REGION_DETECTION           = "fusedRegionDetection";
+        public static final String GEOCODE                          = "geocode";
```
[the remark this answers is not in the document]
Comment (bdm310) [reply project/app/src/main/java/org/owntracks/android/support/Preferences.java]: Excellent suggestion.

**Comment by bdm310** [comment]
I had considered doing a noop geocoder but I'm not sure why I decided not to. It's in there now along with variable name changes.

**Comment by growse** [comment]
Thanks - I'm mildly allergic to null and depending on things being null, hence why the noop geocoder I think is a clearer implementation.

I'll have a test tonight once I get near a device that can actually run the emulator, but looks good from me.

File: project/app/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -43,6 +43,8 @@
     <string name="preferencesAutostart">Autostart</string>
     <string name="preferencesRemoteCommand">Remote commands</string>
     <string name="preferencesRemoteCommandSummary">Enable remote commands</string>
+    <string name="preferencesGeocode">Geocode</string>
+    <string name="preferencesGeocodeSummary">Enable geocoding</string>
```
Comment (linusg) [inline project/app/src/main/res/values/strings.xml]: I can contribute German translations for this:

```xml
    <string name="preferencesGeocode">Geocodierung</string>
    <string name="preferencesGeocodeSummary">Geocodierung aktivieren</string>
```

this follows the translations for e.g. remote commands:

https://github.com/owntracks/android/blob/d2e1c22c927927e9dfd8f64a8171188e4d06ffa9/project/app/src/main/res/values-de/strings.xml#L36-L37

**Comment by bdm310** [comment]
I feel like I know Spanish well enough for this. Added both translations.

**Comment by jpmens** [comment]
@bdm310 `Habilitar` sounds a bit stiff in that context to me (but it is precise). I personally would have used `Permitir` which means to allow, permit, or _enable_. :-)

**Comment by bdm310** [comment]
While I'd agree, I figured consistency in following the remote commands switch was more desirable.

**Comment by juzam** [comment]
> 
> 
> @bdm310 `Habilitar` sounds a bit stiff in that context to me (but it is precise). I personally would have used `Permitir` which means to allow, permit, or _enable_. :-)

That's very similar to the Italian counterpart `Abilita` and `Permetti`  so it's `Abilita geolocalizzazione` or ` Permetti geolocalizzazione` 

In my own language I think I'd like better the `Abilita` option but it's a matter of taste, as already noted.

**Comment by jpmens** [comment]
Apropos, @juzam, do we already have an Italian translation? :-D

**Comment by juzam** [comment]
> 
> 
> Apropos, @juzam, do we already have an Italian translation? :-D

ha! I thought about that just after sending my previous comment. I can contribute one, and even if sub-optimal, that could be a nice one to have. I don't recall if it's just a matter of translating [strings.xml](https://github.com/owntracks/android/blob/master/project/app/src/main/res/values/strings.xml) and put it in the relevant `vaules-it` directory.

**Comment by jpmens** [comment]
@growse or @binarybucks could you please give @juzam instructions on what needs to be done for a new language pack?

**Comment by growse** [comment]
> > Apropos, @juzam, do we already have an Italian translation? :-D
> 
> ha! I thought about that just after sending my previous comment. I can contribute one, and even if sub-optimal, that could be a nice one to have. I don't recall if it's just a matter of translating [strings.xml](https://github.com/owntracks/android/blob/master/project/app/src/main/res/values/strings.xml) and put it in the relevant `vaules-it` directory.

Roughly, this is right. Like you say, I think we just need a `strings.xml` inside a `values-it` folder in the resource tree, and I think also an update to`resConfigs` in `build.gradle`.

Would you be happy to raise a PR including these?

**Comment by growse** [comment]
So I did a little bit of testing. There's a problem in that the toggle only takes effect on app restart, because `GeocodingProvider` is only instantiated once by Dagger, on application startup. Providing a configuration toggle to users that doesn't do anything until the app is restarted isn't ideal.

So I think the right answer here is that `GeocodingProvider` needs to allow the `geocoder` implementation to be changed dynamically, and that then needs to be wired up to the preferences changing so that if the user toggles the switch, a new implementation is used (and probably the cache cleared).

I'm aware that this issue also applies to the Opencage setting, so it'd be good to fix this problem for that case too.

**Comment by bdm310** [comment]
That was a limitation I left in consciously. Agreed that the desired functionality is that the geocoder gets rebuilt on settings changes but there are already a few settings in there that require a full app restart so I wasn't too worried. If I'm honest, I'm unlikely to get around to implementing what we want any time in the next few weeks. It may be best to merge this now and file bugs to fix up all the settings that need a restart later. If that's not how you want to proceed, we can leave this open and I'll get to it when I have the time.

**Comment by growse** [comment]
Ok, fair enough. I think it's reasonable that we get this change in now. 

Only outstanding request is to update the setting summary to make it clear it requires a restart.

File: project/app/src/main/res/values/strings.xml (hunk 2 above)
Comment (growse) [inline project/app/src/main/res/values/strings.xml]: Maybe

```suggestion
    <string name="preferencesGeocodeSummary">Enable geocoding (requires restart)</string>
```

File: project/app/src/main/res/values-es/strings.xml [hunk 3]
```diff
@@ -39,6 +39,8 @@
     <string name="preferencesAutostart">Autoarranque</string>
     <string name="preferencesRemoteCommand">Comandos remotos</string>
     <string name="preferencesRemoteCommandSummary">Habilitar comandos remotos</string>
+    <string name="preferencesGeocode">Geocodificación</string>
+    <string name="preferencesGeocodeSummary">Habilitar geocodificación</string>
```
Comment (growse) [inline project/app/src/main/res/values-es/strings.xml]: (whatever the "requires restart" equivalent is here).

File: project/app/src/main/res/values-de/strings.xml [hunk 4]
```diff
@@ -35,6 +35,8 @@
     <string name="preferencesAutostart">Autostart</string>
     <string name="preferencesRemoteCommand">Fernkommandos</string>
     <string name="preferencesRemoteCommandSummary">Fernkommandos aktivieren</string>
+    <string name="preferencesGeocode">Geocodierung</string>
+    <string name="preferencesGeocodeSummary">Geocodierung aktivieren</string>
```
Comment (growse) [inline project/app/src/main/res/values-de/strings.xml]: (whatever the "requires restart" equivalent is here)

File: project/app/src/main/java/org/owntracks/android/support/GeocodingProvider.java [hunk 5]
```diff
@@ -30,10 +30,15 @@
     @Inject
     public GeocodingProvider(@AppContext Context context, Preferences preferences) {
         cache = new LruCache<>(40);
-        if("".equals(preferences.getOpenCageGeocoderApiKey())) {
-            geocoder = new GeocoderGoogle(context);
-        } else {
-            geocoder = new GeocoderOpencage(preferences.getOpenCageGeocoderApiKey());
+        if(true == preferences.getGeocodeEnabled()) {
```
[the remark this answers is not in the document]
Comment (bdm310) [reply project/app/src/main/java/org/owntracks/android/support/GeocodingProvider.java]: Sorry, my aerospace is showing.

**Comment by bdm310** [comment]
Someone should proofread my Spanish and German translations.

**Comment by jpmens** [comment]
Strings look good to me.

**Pull request merged**